core/src/main/java/io/substrait/expression/FieldReference.java

```diff
@@ -149,7 +149,12 @@ public FieldReference constructOnExpression(Expression expr) {
 
     @Override
     public FieldReference constructOnRoot(Type type) {
-      return FieldReference.newRootStructReference(offset(), type);
+      if (!(type instanceof Type.Struct)) {
+        throw new IllegalArgumentException(
+            "Root type expected to be Type.Struct but is " + type.getClass());
+      }
+      return FieldReference.newRootStructReference(
```

Comment:
Let's add a check on the offset() being within the range of available fields. It would be more helpful to give a more meaningful error here than aioob with a generic message.

Reply:
Added check